# Comparing gene/protein expression profiles to identify differentially
# expressed molecules (DEM), such as differentially expressed genes (DEG)
# or differentially expressed proteins (DEP).
#
# Input datasets for comparison are the raw read counts of mRNA or peptide
# expression levels, measured at different drug-treated conditions as well
# as control condition and generated by RNA-sequencing or protein-spectrum
# experiments,

compareMoleculeExpression <- function(func_dir, repo_dir, config_file, fdrs=c(0.1), read_counts_title="ReadCounts", exp_configs_title="ExpDesigns", deg_title="DEG", top_title="TOP", calc_title="Calc", plot_title="Plots")
{
    # Turn on warning output.
    orig_warn_status <- getOption("warn")
    options(warn=1)

    ####################### Load required function library #######################

    # Load required library
    require(matrixStats)
    require(tools)
    # These R function scripts need to be in the same directory as this main script.
    source(file.path(func_dir, "initializeParameters.R"), local=TRUE)
    source(file.path(func_dir, "readReadCountsDatasets.R"), local=TRUE)
    source(file.path(func_dir, "mergeReadCountsDatasets.R"), local=TRUE)
    source(file.path(func_dir, "adaptReadCountsDatasets.R"), local=TRUE)
    source(file.path(func_dir, "sumExperimentDesign.R"), local=TRUE)
    source(file.path(func_dir, "filterGeneExpSamples.R"), local=TRUE)
    source(file.path(func_dir, "plotHeatMapOfGeneExpDistance.R"), local=TRUE)
    source(file.path(func_dir, "calcExpressionDifference.R"), local=TRUE)

    ############ Initialize various parameters for comparison analysis ############

    # Set data diretories.
    counts_dir <- file.path(repo_dir, "Counts")
    params_dir <- file.path(repo_dir, "Params")
    scripts_dir <- file.path(repo_dir, "Scripts")
    results_dir <- file.path(repo_dir, "Results")
    repo_dirs <- c(counts_dir, params_dir, scripts_dir, results_dir)
    repo_dirs_avail <- dir.exists(repo_dirs)
    if(!all(repo_dirs_avail)) stop(paste(paste0(repo_dirs[!repo_dirs_avail],collapse=" and "), "doesn't exist!"))

    # Initializing various input parameters for differential expression analysis.
    initializeParameters(config_file=config_file, counts_dir=counts_dir, params_dir=params_dir, scripts_dir=scripts_dir)

    ############# Read in read counts data experiment design schemes #############

    # Step 1. Read in read counts datasets and experiment design schemes.
    read_counts_datasets_original <- readReadCountsDatasets(read_counts_files, generateReadCountsColumnNames=generateReadCountsColumnNames, split_char=split_char, field_names=field_names, field_types=field_types, species_default=species_default, subject_default=subject_default, state_default=state_default, culture_default=culture_default, measure_default=measure_default, time_default=time_default, row_name_column=row_name_column, read_counts_columns=read_counts_columns, read_counts_skip=read_counts_skip, exprt_design_files=exprt_design_files, field_names_positions=field_names_positions, exprt_design_skip=exprt_design_skip, func_dir=func_dir)

    # Step 2. Perform gene/protein-specific processing of read counts datasets.
    if(!is.null(processReadCountsDatasets))
    {
        processReadCountsDatasets <- match.fun(processReadCountsDatasets)
        stopifnot(is.function(processReadCountsDatasets))
        read_counts_datasets_original <- processReadCountsDatasets(read_counts_datasets_original)
    }

    # Step 3. Merge multiple read counts matrices and experiment design tables.
    # Note: read_counts_datasets_merged may contain all-zero rows.
    read_counts_datasets_merged <- mergeReadCountsDatasets(read_counts_datasets=read_counts_datasets_original, merge_method=read_counts_merge_method, missing_value=missing_value, func_dir=func_dir)

    # Assign merged read counts matrix and experiment design table to separate
    # variables for compatibility with subsequent comparison analysis.
    read_counts_merged <- read_counts_datasets_merged$read_counts
    targets_merged <- read_counts_datasets_merged$exprt_design

    # Step 4. Adapt read counts data fields to the format used in comparison analysis.
    # Adaptation rules:
    # Subject --> Cell
    # Culture --> Experiment
    # Replicate --> Dish
    # Measure --> Plate
    # Sample --> Well
    name_map <- c(Subject="Cell", Culture="Experiment", Replicate="Dish", Measure="Plate", Sample="Well")
    targets_merged <- adaptReadCountsDatasets(targets_merged, name_map=name_map)
    # Adjust the order to experiment design fields.
    targets_merged <- targets_merged[,c("Time", "Species", "Cell", "State", "Experiment", "Dish", "Plate", "Well", "ID")]
    # Save modified experiment design table.
    read_counts_datasets_merged$exprt_design <- targets_merged

    # Misc data customization.
    # Change all CTRL-related state name to CTRL.
    targets_merged$State[grepl(state_name_ctrl, targets_merged$State, ignore.case=TRUE)] <- state_name_ctrl
    # Sort targets_merged by Cell, State, and Experiment.
    targets_merged <- targets_merged[order(xtfrm(targets_merged[,"Species"]),xtfrm(targets_merged[,"Cell"]),xtfrm(targets_merged[,"State"]),xtfrm(targets_merged[,"Experiment"]),xtfrm(targets_merged[,"Dish"]),xtfrm(targets_merged[,"Plate"]),xtfrm(targets_merged[,"Well"])),]

    ############# Post process read counts data experiment design schemes #############

    # Set plate number to 0 for merging sample replicates across different plates.
    if(merge_plates)
    {
        targets_merged$Plate.Orig <- targets_merged$Plate
        targets_merged$Plate <- merged_plate_number
        targets_merged <- targets_merged[,c("Time", "Species", "Cell", "State", "Experiment", "Dish", "Plate", "Plate.Orig", "Well", "ID")]
    }

    # Set the name of cell species and time points.
    # Only allow single cell species and time point in differential expression
    # comparison, because the compariosns involving multiple cell species and
    # time points needs to be set up in different ways.
    bio_spec <- unique(targets_merged$Species)
    stopifnot(length(bio_spec)==1)
    time_point <- unique(targets_merged$Time)
    stopifnot(length(time_point)==1)
    time_length <- paste("Hour", time_point, sep=".")

    # Save the filtered targets and read counts.
    targets_merged_file <- file.path(results_dir, "Experiment-Design-Merged.tsv")
    write.table(targets_merged, targets_merged_file, quote=FALSE, sep="\t", row.names=FALSE)
    # Write the read counts to a data file.
    read_counts_merged_file <- file.path(results_dir, "Read-Counts-Merged.tsv")
    write.table(read_counts_merged, read_counts_merged_file, quote=FALSE, sep="\t")

    #################### Summarize original experiment design ####################

    state_cell_plate_rep <- sumExperimentDesign(targets_merged)
    if(!is.null(state_cell_plate_rep))
    {
        drug_sample_reps <- data.frame(No=rownames(state_cell_plate_rep), state_cell_plate_rep, stringsAsFactors=FALSE)
        colnames(drug_sample_reps) <- c("No", "Drug", "Cell", "Plate", "Rep")
        drug_sample_reps_file <- file.path(results_dir, "Drug-Sample-Reps-Original.tsv")
        write.table(drug_sample_reps, drug_sample_reps_file, sep="\t", quote=FALSE, row.names=FALSE)
        if(verbose1) print("Original Drug Sample Replicates:")
        if(verbose1) print(state_cell_plate_rep)
    }

    ####################### Read in pre-defined drug groups #######################

    # The file of drug groups represents a global drug-group config for
    # all merged experiments.
    drug_groups_table <- read.delim(file=drug_groups_file, quote="", comment.char="#", stringsAsFactors=FALSE)
    drug_groups <- list()
    drug_groups[[state_name_ctrl]] <- state_name_ctrl
    # Assert no duplicated drug names in drug_groups.
    stopifnot(all(!duplicated(drug_groups_table$Drug)))
    for(drug_group in drug_groups_table$Group) drug_groups[[drug_group]] <- drug_groups_table[drug_groups_table$Group==drug_group,]$Drug

    # Compare the pre-defined drug groups with available treatment conditions
    # in targets (experiment design) to detech if there are any mismatches. And
    # it's possible that not all pre-defined drug names of each drug group can be
    # matched with available drug-treated conditions in each cell line.
    drug_names <- unique(unlist(drug_groups, use.names=FALSE))
    state_names <- unique(targets_merged$State)
    if(verbose1) print(paste("Pre-defined drug groups and available experiment conditions are", (if(setequal(drug_names, state_names)) "the same" else "different")))
    diff_drug_state <- setdiff(drug_names, state_names)
    if(length(diff_drug_state) > 0)
    {
        if(verbose1) print("The drugs defined in drug groups but unavaible in experiment conditions:")
        if(verbose1) print(diff_drug_state)
    }
    diff_state_drug <- setdiff(state_names, drug_names)
    if(length(diff_state_drug) > 0)
    {
        if(verbose1) print("The drugs avaible in experiment conditions but undefined in drug groups:")
        if(verbose1) print(diff_state_drug)
    }

    ###################### Remove outlier sample replicates ######################

    # Filter out outlier samples according to divisive clustering analysis, such that
    # any samples which depart from the rest samples by greater than dist_cutoff_outlier
    # (for group with more than min_samples number of samples) or dist_cutoff_group are
    # removed (for group with min_samples number of samples).

    # Load customized cutoff values for outlier samples.
    outlier_cutoffs <- read.delim(file=outlier_cutoffs_file, quote="", comment.char="#", stringsAsFactors=FALSE)

    # Prepare PDF file for plotting.
    if(plot_clust)
    {
        clust_main_title <- paste(bio_spec, time_length, sep="-")
        clust_plot_file <- file.path(results_dir, paste(paste(clust_main_title, "Cluster-Plots", sep="-"), "pdf", sep="."))
        pdf(file=clust_plot_file, width=8.5, height=11, onefile=TRUE)
        orig_par <- par(no.readonly = TRUE)
        # Plot layout: 3 rows and 2 columns.
        par(mfrow=c(3,2))
    }

    # Sort and filter sample replicates in each drug-treated condition for
    # all drug groups and cell lines.
    drug_names <- unique(unlist(drug_groups, use.names=FALSE))
    targets_merged <- filterGeneExpSamples(targets_merged=targets_merged, read_counts_merged=read_counts_merged, drug_names=drug_names, dist_cutoffs=outlier_cutoffs, dist_cutoff_outlier=dist_cutoff_outlier, dist_cutoff_group=dist_cutoff_group, min_samples=min_samples, filter_outlier=TRUE, plot_clust=plot_clust, verbose=(if(verbose1) verbose2 else verbose1), func_dir=func_dir)
    read_counts_merged <- read_counts_merged[,targets_merged$ID]
    # Remove the rows containing all zero read counts due to the removal of outlier samples.
    read_counts_merged <- read_counts_merged[rowSums(read_counts_merged)>0,]
    # Save filtered experiment design and read counts datasets.
    read_counts_datasets_filtered <- list(exprt_design=targets_merged, read_counts=read_counts_merged)

    # Save the filtered targets and read counts.
    targets_filtered_file <- file.path(results_dir, "Experiment-Design-Filtered.tsv")
    write.table(targets_merged, targets_filtered_file, quote=FALSE, sep="\t", row.names=FALSE)
    # Write the read counts to a data file.
    read_counts_filtered_file <- file.path(results_dir, "Read-Counts-Filtered.tsv")
    write.table(read_counts_merged, read_counts_filtered_file, quote=FALSE, sep="\t")

    # Restore orginal graphics settings.
    if(plot_clust)
    {
        par(orig_par)
        # Close the PDF plot.
        dev.off()
    }

    #################### Summarize filtered experiment design ####################

    state_cell_plate_rep_filter <- sumExperimentDesign(targets_merged)
    if(!is.null(state_cell_plate_rep_filter))
    {
        drug_sample_reps <- data.frame(No=rownames(state_cell_plate_rep_filter), state_cell_plate_rep_filter, stringsAsFactors=FALSE)
        colnames(drug_sample_reps) <- c("No", "Drug", "Cell", "Plate", "Rep")
        drug_sample_reps_filter_file <- file.path(results_dir, "Drug-Sample-Reps-Filtered.tsv")
        write.table(drug_sample_reps, drug_sample_reps_filter_file, sep="\t", quote=FALSE, row.names=FALSE)
        if(verbose1) print("Filtered Drug Sample Replicates:")
        if(verbose1) print(state_cell_plate_rep_filter)
    }

    # Check the availability of targets_merged after outlier removal.
    if(is.null(targets_merged) || (!is.null(targets_merged) && nrow(targets_merged)==0)) stop("Experiment design after outlier removal is empty!")

    #################### Plot the heat map of sample distance ####################

    if(plot_heatmap)
    {
        # Prepare PDF file for plotting.
        dist_main_title <- paste(bio_spec, time_length, sep="-")
        dist_plot_file <- file.path(results_dir, paste(paste(dist_main_title, "Distance-Plots", sep="-"), "pdf", sep="."))
        pdf(file=dist_plot_file, width=8.5, height=8.5, onefile=TRUE)

        # Plot the drugs of each groups across all cell lines.
        for(drug_group_name in names(drug_groups))
        {
            drug_group <- drug_groups[[drug_group_name]]
            for(state_name in drug_group)
            {
                if(verbose1) print(paste("Drug", state_name, "in", drug_group_name, "group across cell lines."))
                # Prepare the datasets.
                targets_merged_state <- targets_merged[targets_merged$State==state_name,]
                # Some drugs have been removed by outlier-removal step.
                if(nrow(targets_merged_state) < 2) next
                wellplate_names_state <- targets_merged_state$ID
                state_names_state <- targets_merged_state$State
                cell_names_state <- targets_merged_state$Cell
                read_counts_state <- read_counts_merged[,wellplate_names_state]
                read_counts_state <- read_counts_state[rowSums(is.na(read_counts_state))==0,]
                if(is.vector(read_counts_state)) read_counts_state <- as.matrix(read_counts_state)
                # Plot heat map of distance matrix.
                cell_names_str <- paste0(unique(cell_names_state),collapse="/")
                title_text <- paste(ncol(read_counts_state), state_name, "Samples of", drug_group_name, "Group in Cell", cell_names_str)
                plotHeatMapOfGeneExpDistance(gene_exp=read_counts_state, group_names=state_names_state, use_group_names=FALSE, title_text=title_text, margin_bottom=8, dist_method="correlation", normalize=FALSE, lower_limit=lower_limit, upper_limit=upper_limit, func_dir=func_dir)
            }
        }

        # Close the PDF plot.
        dev.off()
    }

    ####### Compare read counts data and obtain differentially expressed molecules ##########

    # Determine whether or not to generate various data files.
    if(!write_read_counts_file) read_counts_title <- NULL
    if(!write_exp_configs_file) exp_configs_title <- NULL
    if(!write_deg_file) deg_title <- NULL
    if(!write_top_file) top_title <- NULL
    if(!write_calc_file) calc_title <- NULL
    if(!write_plot_file) plot_title <- NULL

    # Compare the difference of molecular expression levels for genes and proteins, and
    # return both raw read counts and differential comparison statistics of all molecules.
    deg_read_counts <- calcExpressionDifference(read_counts=read_counts_merged, exprt_design=targets_merged, drug_groups=drug_groups, bio_spec=bio_spec, time_length=time_length, results_dir=results_dir, fdrs=fdrs, state_name_ctrl=state_name_ctrl, min_samples=min_samples, rep_req=rep_req, dispersion=dispersion, n_top=n_top, deg_only=deg_only, norm_base=norm_base, remove_small=remove_small, plot_bcv=plot_bcv, plot_smear=plot_smear, pt_chars=pt_chars, sub_char=sub_char, read_counts_title=read_counts_title, deg_title=deg_title, top_title=top_title, exp_configs_title=exp_configs_title, calc_title=calc_title, plot_title=plot_title, verbose1=verbose1, verbose2=verbose2, func_dir=func_dir)

    ################################## Clean up ##################################

    # Restore warning output.
    options(warn=orig_warn_status)

    # Save and return multiple datasets of read counts and experiment design.
    # Original: multiple original molecular read counts
    # Merged: the merged molecular read counts from multiple original read counts.
    # Filtered: the filtered molecular read counts after the removal of outlier samples and all-zero read counts.
    read_counts_datasets <- list(original=read_counts_datasets_original, merged=read_counts_datasets_merged, filtered=read_counts_datasets_filtered, degstats=deg_read_counts)
    # Save the dataset object to a RData file.
    if(!is.null(read_counts_datasets_rdata_file))
    {
        if(tolower(file_ext(read_counts_datasets_rdata_file)) != "rdata") read_counts_datasets_rdata_file <- paste(read_counts_datasets_rdata_file, "RData", sep=".")
        read_counts_datasets_rdata_file <- file.path(results_dir, read_counts_datasets_rdata_file)
        save(read_counts_datasets, file=read_counts_datasets_rdata_file)
    }

    return(read_counts_datasets)
}
